src/mycollections/DoubleLinkedList.java

```diff
@@ -1,59 +1,245 @@
+public class DoubleLinkedList<T> implements MyList<T> {
+
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
 
     @Override
     public int size() {
-        return 0;
+        return this.size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
     }
 
     @Override
     public boolean contains(T o) {
-        return false;
+        elementValueCheck(o);
+        return indexOf(o) != -1;
     }
 
     @Override
     public int indexOf(T o) {
-        return 0;
+        elementValueCheck(o);
+        int index = 0;
+        for (Node<T> first = head; first != null; first = first.next) {
+            if (first.item.equals(o)) {
+                return index;
+            }
+            index++;
+        }
+        return -1;
     }
 
     @Override
     public int lastIndexOf(T o) {
-        return 0;
+        elementValueCheck(o);
+        int index = 0;
+        for (Node<T> last = tail; last != null; last = last.prev) {
+            if (last.item.equals(o)) {
+                return index;
+            }
+            index++;
+        }
+        return -1;
     }
 
     @Override
     public T get(int index) {
+        rangeCheck(index);
+        if (size != 0) {
+            int halfLength = (size >> 1);
+            if (index <= halfLength && head != null) {
+                for (int i = 0; i <= halfLength; i++) {
+                    if (i == index) {
+                        return head.item;
+                    }
+                    head = head.next;
+                }
+            } else if (index > halfLength && tail != null) {
+                for (int i = size; i > halfLength; i--) {
+                    if (i == index) {
+                        return tail.item;
+                    }
+                    tail = tail.prev;
+                }
+            }
+        }
         return null;
     }
 
     @Override
     public T set(int index, T element) {
-        return null;
+        rangeCheck(index);
+        elementValueCheck(element);
+        Node<T> elementToSet = new Node<>(null, element, null);
+        Node<T> initialElement = getByIndex(index);
+        if (index == 0) {
+            elementToSet.next = head.next;
+            head = elementToSet;
+        } else if (index == size) {
+            elementToSet.prev = tail.prev;
+            tail = elementToSet;
+        } else {
+            elementToSet.next = initialElement.next;
+            elementToSet.prev = initialElement.prev;
+        }
+        initialElement.next = initialElement.prev = null;
+        return initialElement.item;
     }
 
     @Override
     public boolean add(T e) {
-        return false;
+        elementValueCheck(e);
+        linkLast(e);
+        return true;
     }
 
     @Override
     public void add(int index, T element) {
-
+        rangeCheck(index);
+        elementValueCheck(element);
+        Node<T> elementToAdd = new Node<>(null, element, null);
```

Comment:
Before changing already existing element value, check if (index == size) and if yes call add(E e) method for simple add, otherwise execute your logic (your logic is incorrect, please debug it).